File: examples/zig/if-else/IfElse_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const IfElse = @import("IfElse.runar.zig").IfElse;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "if-else/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check IfElse.runar.zig" {
    try runCompileChecks("IfElse.runar.zig");
}

test "IfElse.check passes on the true branch when value+limit is positive" {
    const c = IfElse.init(5);
    c.check(3, true);
}

test "IfElse.check passes on the false branch when value-limit is positive" {
    const c = IfElse.init(5);
    c.check(20, false);
}

File: examples/ts/if-else/IfElse.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'IfElse.runar.ts'), 'utf8');

describe('IfElse', () => {
  it('compiles via TestContract.fromSource', () => {
    const c = TestContract.fromSource(source, { limit: 5n });
    expect(c.state.limit).toBe(5n);
  });

  it('passes on the true branch (value + limit > 0)', () => {
    const c = TestContract.fromSource(source, { limit: 5n });
    const r = c.call('check', { value: 3n, mode: true });
    expect(r.success).toBe(true);
  });

  it('passes on the false branch (value - limit > 0)', () => {
    const c = TestContract.fromSource(source, { limit: 5n });
    const r = c.call('check', { value: 20n, mode: false });
    expect(r.success).toBe(true);
  });

  it('rejects when neither branch yields a positive result', () => {
    const c = TestContract.fromSource(source, { limit: 5n });
    const r = c.call('check', { value: 1n, mode: false });
    expect(r.success).toBe(false);
  });
});
